satellite/overlay/combinedcache.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package overlay
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+type addressInfo struct {
+	address   string
+	lastIP    string
+	transport pb.NodeTransport
+}
+
+type uptimeInfo struct {
+	isUp       bool
+	lastUptime time.Time
+	stats      *NodeStats
+}
+
+// CombinedCache is a simple caching mechanism for overlaycache updates. It
+// provdes methods to help reduce calls to UpdateAddress and UpdateTime, but can
+// be extended for other calls in the future.
+type CombinedCache struct {
+	DB
+	addressLock  sync.RWMutex
+	addressCache map[storj.NodeID]*addressInfo
+
+	uptimeLock          sync.RWMutex
+	uptimeCache         map[storj.NodeID]*uptimeInfo
+	uptimeFlushInterval time.Duration
+}
+
+// NewCombinedCache instantiates a new CombinedCache
+func NewCombinedCache(db DB, uptimeFlushInterval time.Duration) *CombinedCache {
+	return &CombinedCache{
+		DB:                  db,
+		addressCache:        make(map[storj.NodeID]*addressInfo),
+		uptimeCache:         make(map[storj.NodeID]*uptimeInfo),
+		uptimeFlushInterval: uptimeFlushInterval,
+	}
+}
+
+// UpdateUptime overrides the underlying db.UpdateUptime and provides a simple
+// caching layer to reduce calls to the underlying db.
+func (c *CombinedCache) UpdateUptime(ctx context.Context, nodeID storj.NodeID, isUp bool, lambda, weight, uptimeDQ float64) (stats *NodeStats, err error) {
+	// First check the internal cache. If it returns stats, use them.
+	stats = c.GetNodeStats(nodeID, isUp)
+	if stats != nil {
+		return stats, nil
+	}
+	stats, err = c.DB.UpdateUptime(ctx, nodeID, isUp, lambda, weight, uptimeDQ)
+	if err != nil {
+		return nil, err
+	}
+	// Refresh internal stats
+	c.SetNodeStats(nodeID, isUp, stats)
+	return stats, nil
+}
+
+// UpdateAddress overrides the underlying db.UpdateAddress and provides a simple
+// caching layer to reduce calls to the underlying db.
+func (c *CombinedCache) UpdateAddress(ctx context.Context, info *pb.Node, defaults NodeSelectionConfig) (err error) {
```

Comment:
There seems to be a race condition in calls to UpdateAddress- observations that we make about a node might end up getting passed to UpdateAddress in a different order from the order they happened in. Much of the time that won't matter, but I haven't been able to prove to myself that it will _never_ matter.

Picture this scenario-

* Node N contacts satellite asking for pingback at address A1
* Satellite goroutine G1 gets the request, pings node N, makes a successful connection to A1
* Node N switches to using a different interface with a different IP
* Node N contacts satellite asking for pingback at address A2
* Satellite goroutine G2 gets the request, pings node N, makes a successful connection to A2
* Goroutine G2 calls UpdateAddress, gets the addressCache lock, writes address A2 to cache and db
* Goroutine G1 calls UpdateAddress, gets the addressCache lock, writes address A1 to cache and db

So the addressCache and the db would not actually have the latest contacted IP. Your class invariants would be violated. Does that matter even then? I'm not sure. Maybe not. But if a race condition like this exists, and there is a way to break the described behavior of a software component, it _will_ totally happen because computers are the worst and they hate us. Also it sounds possible for malicious node operators to cause this situation on purpose fairly reliably.

If you're sure that all of that is harmless and acceptable, then that's ok too, but we just need to change some comments and docstrings and possibly method names to make it clear that addressCache might not have the absolute latest address and IP, and the db might not have the absolute latest address and IP, and that instead it's a "best effort" situation in both cases.

Reply:
I believe I have fixed this with my latest update.

Reply:
To really avoid the race between UpdateUptime calls with time-separated connections to the same node, I want to suggest using a channel and a dedicated goroutine to govern writes into the cache. Depending on how reads get managed, you might be able to do away with all the mutexes and locks in this code altogether.

Reply:
We discussed this in person, just recording a few thoughts here. As we discussed, it appears this race already exists -- that is, a badly-behaving node can already request two ping-backs at different addresses such that we can't guarantee the order in which they arrive int he database. My update should not change this behavior, but should guarantee that the db and cache remain in sync (which I believe it does).

It does appear, as we discussed, that these calls all happen synchronously. I can add a monotonic clock check for each UpdateAddress request and toss out any request that is older than a previous request we've seen, though I still believe that this should be handled in a separate ticket and seems outside the scope of my ticket. :smile_cat:

Reply:
And again, I am fine with leaving the race as it is, and filing a ticket or another PR or whatever to address later, but until that happens it is imperative that the weaker semantics be documented. The class can not claim to cache the latest address always- it can just say “probably the latest” or “best effort” or something like that. Our components just need to be able to rely on each other absolutely for us to build a robust and secure distributed system.

Reply:
I think I understand better now. When I say "latest address always" I'm referring to the "last address updated in the database." I see what you mean by the fact that it may not actually be the latest in a real-world-time sense. I will definitely document that, thank you for pointing it out and helping me understand what you mean. Sorry that I've been a little slow on the uptake there to fully understand.